src/global.c

```diff
@@ -387,3 +370,48 @@ git_global_st *git__global_state(void)
 }
 
 #endif /* GIT_THREADS */
+
+#ifdef GIT_WIN32 
+BOOL git__dllmain(HINSTANCE hInstDll, DWORD fdwReason, LPVOID lpvReserved)
+{
+	GIT_UNUSED(hInstDll);
+	GIT_UNUSED(lpvReserved);
+
+#ifdef GIT_THREADS
+	/* This is how Windows lets us know our thread is being shut down */
+	if (fdwReason == DLL_THREAD_DETACH) {
+		git__free_tls_data();
+	}
+#else
+	GIT_UNUSED(fdwReason);
+#endif /* GIT_THREADS */
+
+	/*
+	 * Windows pays attention to this during library loading. We don't do anything
+	 * so we trivially succeed.
+	 */
+	return TRUE;
+}
+
+#ifdef GIT_STATIC
+BOOL git_libgit2_dllmain(HINSTANCE hInstDll, DWORD fdwReason, LPVOID lpvReserved)
+{
+	return git__dllmain(hInstDll, fdwReason, lpvReserved);
+}
+#else
+BOOL git_libgit2_dllmain(HINSTANCE hInstDll, DWORD fdwReason, LPVOID lpvReserved)
+{
+	GIT_UNUSED(hInstDll);
+	GIT_UNUSED(fdwReason);
+	GIT_UNUSED(lpvReserved);
+
+	return TRUE;
+}
```

Comment:
And subsequently this function wouldn't need a definition `#ifndef GIT_STATIC`.

Reply:
This requirement is on purpose. If you do it my way, a user application that uses git_libgit2_dllmain() and supports linking against a static _or_ shared build of libgit2 does not need to handle those cases separately.

Reply:
Agreed